Source/Fuse.Models/JavaScript.Model.uno

```diff
@@ -0,0 +1,266 @@
+					result.Args.Add( argName );
+				}
+
+				return result;
+			}
+			
+			throw new Exception("Invalid Model path expression: " + exp);
+		}
+
+		void SetupModel()
+		{
+			if (_model == null)
+			{
+				Code = string.Empty;
+				return;
+			}
+
+			//TODO: this should not be necessary. It's done because there's a UX processor error, we get
+			//the IExpression prior to it being complete
+			var module = ParseModelExpression( _model.Source, _nameTable );
+
+			Dependencies.Clear();
+			for (int i=0; i < module.Dependencies.Count; ++i)
+				Dependencies.Add( module.Dependencies[i] );
+			
+			var code = "var Model = require('FuseJS/Internal/Model');\n"+
+					"var ViewModelAdapter = require('FuseJS/Internal/ViewModelAdapter')\n";
+					
+			code += "var self = this;\n"+
+					"var modelClass = require('" + module.ModuleName + "');\n"+
+					"if (!(modelClass instanceof Function) && 'default' in modelClass) { modelClass = modelClass.default }\n"+
+					"if (!(modelClass instanceof Function) && '" + module.ClassName +"' in modelClass) { modelClass = modelClass."+ module.ClassName +" }\n"+
+					"if (!(modelClass instanceof Function)) { throw new Error('\"" + module.ModuleName + "\" does not export a class or function required to construct a Model'); }\n"+
+					"var modelInstance = Object.create(modelClass.prototype);\n"+
+					"module.exports = new Model(modelInstance, function() {\n"+
+					"    modelClass.call(modelInstance" + module.ArgString + ");\n"+
+					"    ViewModelAdapter.adaptView(self, module, modelInstance);\n"+
+					"    return modelInstance;\n"+
+					"});\n";
+			Code = code;
+		}
+		
+		string _previewStateModelId; //if null then not migrated
+		
+		static public ModelJavaScript CreateFromPreviewState(Visual where, IExpression model)
+		{
+			string previewStateId = "ModelJavaScript-App";
+			var parsed = ParseModelExpression(model, null);
+			
+			var previewState = PreviewState.Find(where);
+			if (previewState != null && previewState.Current != null)
+			{
+				var previous = previewState.Current.Consume( previewStateId ) as ModelJavaScript;
+				if (previous != null)
+				{
+					if (previous._model.CompatibleTo(parsed))
+						return previous;
+					else
+						previous.Dispose();
+				}
+			}
+			
+			//app-level model does not have a nametable otherwise migration would not be possible
+			var js = new ModelJavaScript(parsed, null, previewStateId );
+			return js;
+		}
+		
+		ParsedModelExpression _model;
+		internal ModelJavaScript(ParsedModelExpression model, NameTable nt,
+			string previewStateId)
+			: base(nt)
+		{
+			_previewStateModelId = previewStateId;
+			_model = model;
+			FileName = "(model-script)";
+		}
+		
+		protected override void OnRooted()
+		{
+			//TODO: prior to base.OnRooted is questionable... The TODO: in SetupModel though is part of the
+			//reason we can't easily fix this now. Ideally the constructor would just set the needed values
```

Comment:
Should this be investigated? If not, please add a ticket to do so, and link it here...